keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,1338 @@
+
+Currently after CreateVolume with provider specific parameters pass in storage class, Kubernetes does not handle update of volume attributes other than capacity. The only way to achieve this is to use the cloud provider's API directly to modify iops and throughput. This creates frictions of managing storage resources on k8s. Allowing Resources to be changed without managing it through different provider's APIs directly simplifies this whole process.
+
+### Goals
+
+* Cloud-provider and storage vendor independent specification of basic volume attributes parameters
+* Volume attributes enforcement performed by the storage
+* Specifying volume attributes at volume creation
+* Changing volume attributes parameters on-line for existing volumes
+* Volume attributes modifiable non disruptively by workload developers
+* ResourceQuota enforcement by VolumeAttributesClass
+
+### Non-Goals
+
+* OS-level IO volume attributes, for example cgroup policy
+* Inter-pod volume attributes. This KEP is restricted to what attributes can be defined on a volume.
+* Scheduling based on total specific volume attributes limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+We need to add a new resource object VolumeAttributesClass to Kubernetes API, also a new admission controller and vac protection controller. Please see more in [Design Details](#bookmark=id.wtvwymf8202g).
+
+
+```
+type VolumeAttributesClass struct {
+  metav1.TypeMeta
+  // +optional
+  metav1.ObjectMeta
+
+  // Spec defines the behavior of a VolumeAttributesClass.
+  // +optional
+  Spec VolumeAttributesClassSpec
+}
+
+type VolumeAttributesClassSpec struct {
+  // Name of the drive
+  DriverName string
+  // This field is OPTIONAL.This allows the CO to specify the
+  // volume attributes class parameters to apply. 
+  Parameters map[string]string
+}
+
+type PersistentVolumeSpec struct {
+  ...
+  // Name of VolumeAttributesClass to which this persistent volume belongs. Empty value
+  // means that this volume does not belong to any VolumeAttributesClass.
+  // +optional
+  VolumeAttributesClassName string
+  ...
+}
+
+type PersistentVolumeClaimSpec struct {
+  ...
+  // Name of the VolumeAttributesClass required by the claim.
+  // More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1
+  // +optional
+  VolumeAttributesClassName *string
+  ...
+}
+
+// PersistentVolumeClaimStatus represents the status of PV claim
+type PersistentVolumeClaimStatus struct {
+  ...
+  VolumeAttributesClassName string
```

Comment:
I'm not sure I fully understand the relationship between `PersistentVolumeSpec.VolumeAttributesClassName`, `PersistentVolumeClaimSpec.VolumeAttributesClassName` and `PersistentVolumeClaimStatus.VolumeAttributesClassName`.

I understand from the examples how `PersistentVolumeClaimSpec.VolumeAttributesClassName` is used. What are the other two for?

Reply:
So we need to add PVSpec is for the preprovisioned volume scenario to the design below?